src/Cli/dotnet/commands/dotnet-test/MSBuildUtility.cs

```diff
@@ -23,18 +24,70 @@ public static (IEnumerable<Module> Projects, bool IsBuiltOrRestored) GetProjects
                     SolutionAndProjectUtility.GetRootDirectory(solutionFilePath);
 
             ConcurrentBag<Module> projects = GetProjectsProperties(new ProjectCollection(), solutionModel.SolutionProjects.Select(p => Path.Combine(rootDirectory, p.FilePath)), buildOptions);
+
+            isBuiltOrRestored |= !projects.IsEmpty;
+
             return (projects, isBuiltOrRestored);
         }
 
         public static (IEnumerable<Module> Projects, bool IsBuiltOrRestored) GetProjectsFromProject(string projectFilePath, BuildOptions buildOptions)
         {
             bool isBuiltOrRestored = BuildOrRestoreProjectOrSolution(projectFilePath, buildOptions);
 
-            IEnumerable<Module> projects = SolutionAndProjectUtility.GetProjectProperties(projectFilePath, GetGlobalProperties(buildOptions), new ProjectCollection());
+            IEnumerable<Module> projects = SolutionAndProjectUtility.GetProjectProperties(projectFilePath, GetGlobalProperties(buildOptions.BuildProperties), new ProjectCollection());
+
+            isBuiltOrRestored |= projects.Any();
 
             return (projects, isBuiltOrRestored);
         }
 
+        public static BuildOptions GetBuildOptions(ParseResult parseResult, int degreeOfParallelism)
+        {
+            IEnumerable<string> propertyTokens = GetPropertyTokens(parseResult.UnmatchedTokens);
+            IEnumerable<string> binaryLoggerTokens = GetBinaryLoggerTokens(parseResult.UnmatchedTokens);
+
+            var msbuildArgs = parseResult.OptionValuesToBeForwarded(TestCommandParser.GetCommand())
+                .Concat(propertyTokens)
+                .Concat(binaryLoggerTokens).ToList();
+
+            List<string> unmatchedTokens = [.. parseResult.UnmatchedTokens];
+            unmatchedTokens.RemoveAll(arg => propertyTokens.Contains(arg));
+            unmatchedTokens.RemoveAll(arg => binaryLoggerTokens.Contains(arg));
+
+            PathOptions pathOptions = new(parseResult.GetValue(TestingPlatformOptions.ProjectOption),
+                parseResult.GetValue(TestingPlatformOptions.SolutionOption),
+                parseResult.GetValue(TestingPlatformOptions.DirectoryOption));
```

Comment:
Nit: for symmetry with the declaration below.

```suggestion
            PathOptions pathOptions = new(parseResult.GetValue(
                TestingPlatformOptions.ProjectOption),
                parseResult.GetValue(TestingPlatformOptions.SolutionOption),
                parseResult.GetValue(TestingPlatformOptions.DirectoryOption));
```